crates/chain/src/chain.rs

```diff
@@ -217,85 +248,337 @@ impl Chain {
             let status: u8 = self.local_state.last_global_state.status().into();
             Status::try_from(status).expect("invalid status")
         };
-        let event = match (status, context) {
-            (Status::Running, L1ActionContext::SubmitTxs { deposit_requests }) => {
-                // Submit transactions
-                // parse layer2 block
-                let l2block = parse_l2block(&transaction, &self.rollup_type_script_hash)?;
-                let number: u64 = l2block.raw().number().unpack();
-                if let Some(challenge_context) = self.process_block(
-                    db,
-                    l2block,
-                    l2block_committed_info.clone(),
-                    global_state.clone(),
-                    deposit_requests,
-                )? {
-                    // stop syncing and return event
-                    self.bad_block_context = Some(challenge_context.target.clone());
-                    SyncEvent::BadBlock(challenge_context)
-                } else {
-                    log::info!("sync new block #{} success", number);
-                    SyncEvent::Success
+
+        let update = || -> Result<SyncEvent> {
+            match (status, context) {
+                (
+                    Status::Running,
+                    L1ActionContext::SubmitBlock {
+                        l2block,
+                        deposit_requests,
+                        reverted_block_hashes,
+                    },
+                ) => {
+                    // Clear reverted l2blocks. It doesn't matter current l2block is bad or not, since
+                    // custodian and withdrawal are reverted on chain.
+                    self.pending_revert_blocks
+                        .retain(|block| !reverted_block_hashes.contains(&block.hash()));
```

Comment:
Because they are included in new l2block's rollup action on-chain. Withdrawals and custodians included in these blocks are also reverted on-chain.

Blocks in ```pending_revert_blocks``` will be packaged in next new l2 block's rollup action. If we don't remove them, we can't produce next block. Because we can't revert withdrawals and custodians twice.